### ISSUE TITLE
more than one redundant closure causes E0382: use of moved value

### ISSUE THREAD

**Issue by Ben-PH** [rust-lang/rust-clippy#7818]
Lint name: `redundant_closure

I tried this code:

```rust
match (left.binary_search_by(|t| cmp_fn), right.binary_search_by(cmp_fn)) {
```
I got this suggestion:

```
warning: redundant closure
   --> table/src/index/rotated_array_set.rs:875:34
    |
875 |                 match (left.binary_search_by(|t| cmp_fn(t)), right.binary_search_by(cmp_fn)) {
    |                                              ^^^^^^^^^^^^^ help: replace the closure with the function itself: `cmp_fn`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```
I expected to see this happen: apply the lint suggestion, and warning to go away

Instead, this happened: 
```
error[E0382]: use of moved value: `cmp_fn`
   --> table/src/index/rotated_array_set.rs:875:66
    |
815 |     fn find_raw_index_by(&self, mut cmp_fn: impl FnMut(&T) -> Ordering) -> Result<usize, usize> {
    |                                 ---------- move occurs because `cmp_fn` has type `impl FnMut(&T) -> Ordering`, which does not implement the `Copy` trait
...
875 |                 match (left.binary_search_by(cmp_fn), right.binary_search_by(cmp_fn)) {
    |                                              ------ value moved here         ^^^^^^ value used here after move
    |
help: consider further restricting this bound
    |
815 |     fn find_raw_index_by(&self, mut cmp_fn: impl FnMut(&T) -> Ordering + Copy) -> Result<usize, usize> {
    |   
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (a8f2463c6 2021-10-09)
binary: rustc
commit-hash: a8f2463c68a6532d74a13ec402ec5b513e4e2726
commit-date: 2021-10-09
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

@ rustbot label +I-suggestion-causes-error

**Comment by Ben-PH** [comment]
The error from the suggested fix can be resolved with `&mut $fn` instead of just `$fn`

**Comment by y21** [comment]
This was fixed at some point and the lint correctly suggests `&mut cmp_fn` now
```
warning: redundant closure
 --> src/main.rs:3:34
  |
3 |     match (left.binary_search_by(|t| cmp_fn(t)), right.binary_search_by(cmp_fn)) {
  |                                  ^^^^^^^^^^^^^ help: replace the closure with the function itself: `&mut cmp_fn`
  |
```
